### PR TITLE
Use HdrHistogram as the default Quantiles implementation

### LINKED ISSUE #793: Failed to handle request by logging listener error
When I upgrade armeria to 0.54.0, maybe because it has some bug on metrics collection, it causes thrift handler failure.

```
2017-10-24 11:06:04.129  WARN 66691 --- [armeria-common-worker-epoll-3-29] c.l.a.c.l.RequestLogListenerInvoker      : onRequestLog() failed with an exception:
java.lang.ArrayIndexOutOfBoundsException: 500
    at io.micrometer.core.instrument.stats.quantile.CKMSQuantiles.observe(CKMSQuantiles.java:81)
    at io.micrometer.core.instrument.prometheus.internal.CustomPrometheusSummary$Child.observe(CustomPrometheusSummary.java:132)
    at io.micrometer.core.instrument.prometheus.PrometheusTimer.record(PrometheusTimer.java:40)
    at com.linecorp.armeria.internal.metric.micrometer.ConcurrentCompositeTimer.lambda$record$0(ConcurrentCompositeTimer.java:67)
    at java.util.concurrent.ConcurrentHashMap$ValuesView.forEach(ConcurrentHashMap.java:4707)
    at com.linecorp.armeria.internal.metric.micrometer.ConcurrentCompositeTimer.record(ConcurrentCompositeTimer.java:67)
    at com.linecorp.armeria.internal.metric.RequestMetricSupport.onResponse(RequestMetricSupport.java:93)
    at com.linecorp.armeria.common.logging.RequestLogListenerInvoker.invokeOnRequestLog(RequestLogListenerInvoker.java:39)
    at com.linecorp.armeria.common.logging.DefaultRequestLog.notifyListeners(DefaultRequestLog.java:607)
    at com.linecorp.armeria.common.logging.DefaultRequestLog.updateAvailability(DefaultRequestLog.java:565)
    at com.linecorp.armeria.common.logging.DefaultRequestLog.updateAvailability(DefaultRequestLog.java:552)
    at com.linecorp.armeria.common.logging.DefaultRequestLog.responseContent(DefaultRequestLog.java:500)
    at com.linecorp.armeria.client.thrift.THttpClientDelegate.handleSuccess(THttpClientDelegate.java:252)
    at com.linecorp.armeria.client.thrift.THttpClientDelegate.handle(THttpClientDelegate.java:219)
    at com.linecorp.armeria.client.thrift.THttpClientDelegate.lambda$execute$0(THttpClientDelegate.java:144)
    at com.linecorp.armeria.common.util.Functions.lambda$voidFunction$1(Functions.java:192)
    at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
    at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
    at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
    at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
    at com.linecorp.armeria.common.HttpMessageAggregator.accept(HttpMessageAggregator.java:120)
    at com.linecorp.armeria.common.HttpMessageAggregator.accept(HttpMessageAggregator.java:29)
    at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
    at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
    at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
    at com.linecorp.armeria.common.stream.DefaultStreamMessage$CloseEvent.notifySubscriber(DefaultStreamMessage.java:633)
    at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriberWithCloseEvent(DefaultStreamMessage.java:403)
    at com.linecorp.armeria.common.stream.DefaultStreamMessage.notifySubscriber(DefaultStreamMessage.java:317)
    at com.linecorp.armeria.common.stream.DefaultStreamMessage.lambda$notifySubscriber$2(DefaultStreamMessage.java:298)
    at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
    at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
    at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:309)
    at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
    at java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

The Quantiles implementations provided by Micrometer are not
thread-safe, causing various races.

Modifications:

- Replace CKMSQuantiles with RollingHdrQuantiles, which is a forked
  version of Rolling Metrics ResetByChunksAccumulator
  - https://github.com/vladimir-bukhtoyarov/rolling-metrics/blob/2.0.3/src/main/java/com/github/rollingmetrics/histogram/accumulator/ResetByChunksAccumulator.java
  - Rolling Metrics supports only long integer values so I had to fork
    it to support double values.
- Add the metrics for HdrHistograms used by RollingHdrQuantiles
- Add MicrometerUtil.registerLater()

Result:

- Fixes #793

### CODE REVIEW

**Comment by trustin** [comment]
Also ported the test case so that `RollingHdrQuantiles` has 95+% test coverage.

File: core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+                        int numberHistoryChunks, long intervalBetweenResettingMillis,
+                        LongSupplier clock) {
+
+        this.intervalBetweenResettingMillis = intervalBetweenResettingMillis;
+        this.clock = clock;
+        creationTimestamp = clock.getAsLong();
+
+        left = new Phase(recorderSupplier, creationTimestamp + intervalBetweenResettingMillis);
+        right = new Phase(recorderSupplier, Long.MAX_VALUE);
+        phases = new Phase[] { left, right};
```
Comment (imasahiro) [inline core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java]: nit. add whitespace before `}`

File: thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java [hunk 2]
```diff
@@ -140,6 +143,7 @@ private static void hello_first_endpoint() throws Exception {
                           "client_total_duration_seconds_sum{handler=\"Foo\","));
 
         final String content = makeMetricsRequest().content().toStringUtf8();
+        logger.debug("Metrics reported by the exposition service:\n{}", content);
```
Comment (imasahiro) [inline thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java]: Not sure but can we remove this debug print?

Comment (trustin) [reply thrift/src/test/java/com/linecorp/armeria/it/metric/PrometheusMetricsIntegrationTest.java]: We can, but it's useful for debugging. Gradle will not print it as long as the test passes.

File: core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java (hunk 1 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java]: Oh, Checkstyle missing it. Fixed :-)

**Comment by trustin** [comment]
Added a gauge that represents the estimated memory footprint of `HdrHistogram`s used by `RollingHdrQuantiles`.

**Comment by trustin** [comment]
Encountered an interesting stuff while working on this PR: https://bugs.openjdk.java.net/browse/JDK-8062841

File: core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+import io.micrometer.core.instrument.stats.quantile.Quantiles;
+
+final class RollingHdrQuantiles implements Quantiles {
+
+    private static final double[] PREDEFINED_QUANTILES = { 0, 0.5, 0.75, 0.9, 0.95, 0.98, 0.99, 0.999, 1.0 };
+    private static final Collection<Double> MONITORED;
+
+    static {
+        final ImmutableList.Builder<Double> builder = ImmutableList.builder();
```
Comment (imasahiro) [inline core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java]: nit. Think we can put list directly than making builder.

```java
MONITORED = Arrays.stream(PREDEFINED_QUANTILES).boxed()
                  .collect(toImmutableList());
```

Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java]: Great idea. Fixed!

File: core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+               .append(array[i]);
+        }
+        buf.append("\n}");
+        return buf.toString();
+    }
+
+    private static final class ArchivedHistogram {
+
+        private final DoubleHistogram histogram;
+        private volatile long proposedInvalidationTimestamp;
```
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/metric/RollingHdrQuantiles.java]: Fixed :-)

**Comment by trustin** [comment]
Thank you, reviewers.

**Pull request merged**